# Homework \#4 - Due Nov. 4
## Hong Tan

In [31]:
import math
import numpy as np

## Chapter 3.2

1. Use Neville’s method to obtain the approximations for Lagrange interpolating polynomials of degrees one, two, and three to approximate each of the following:

In [32]:
# Neville function from professor
def neville(x, vec_x, vec_f, Q_table = None, i0 = -1, j0 = -1):
    n = np.size(vec_x) - 1; # x0, x1, ..., xn.
    
    if (Q_table == None):
        Q_table = np.zeros((n + 1, n + 1));
        
    for i in np.arange(0, n + 1):
        Q_table[i][0] = vec_f[i];

    for j in np.arange(1, n + 1):
        for i in np.arange(j, n + 1):
            # compute Q_{i,j}
            Q_table[i][j] = 0.0;
            Q_table[i][j] += (x - vec_x[i - j])*Q_table[i][j - 1];
            Q_table[i][j] -= (x - vec_x[i])*Q_table[i - 1][j - 1];
            Q_table[i][j] /= (vec_x[i] - vec_x[i - j]);

    return Q_table[n][n], Q_table; 

#### 1.a. $f (8.4)$ if $f (8.1) = 16.94410, f (8.3) = 17.56492, f (8.6) = 18.50515, f (8.7) = 18.82091$.

In [33]:
neville(8.4,[8.1,8.3,8.6,8.7],[16.99410,17.56492,18.50515,18.82091])

(17.8721425,
 array([[16.9941   ,  0.       ,  0.       ,  0.       ],
        [17.56492  , 17.85033  ,  0.       ,  0.       ],
        [18.50515  , 17.87833  , 17.86713  ,  0.       ],
        [18.82091  , 17.87363  , 17.877155 , 17.8721425]]))

Solution: $P_1(8.4) = 17.85033, P_2(8.4)=17.86713, P_3(8.4)=17.8721425$.

#### 1.b. $f(-\frac{1}{3})$ if $f (-0.75) = -0.07181250, f (0.5) = 0.02475000, f (0.25) = 0.33493750,
f (0) = 1.10100000$.

In [34]:
neville(-1/3, [-0.75,0.5,0.25,0], [-0.07181250,0.02475000,0.33493750,1.10100000])

(1.7588703703703699,
 array([[-0.0718125 ,  0.        ,  0.        ,  0.        ],
        [ 0.02475   , -0.039625  ,  0.        ,  0.        ],
        [ 0.3349375 ,  1.05870833,  0.41801389,  0.        ],
        [ 1.101     ,  2.12241667,  2.83155556,  1.75887037]]))

Solution: $P_1(-\frac{1}{3}) = -0.039625, P_2(-\frac{1}{3})=0.41801389, P_3(-\frac{1}{3})=1.75887037$.

#### 3. Use Neville’s method to approximate $\sqrt{3}$ with the following functions and values.
#### (a) $f (x) = 3^x$ and the values $x_0 = -2, x_1 = -1, x_2 = 0, x_3 = 1,$ and $x_4 = 2$.

In [39]:
def f(x):
    return 3**x

neville(math.sqrt(3),[-2,-1,0,1,2],[f(-2),f(-1),f(0),f(1),f(2)])

(6.78024599976228,
 array([[0.11111111, 0.        , 0.        , 0.        , 0.        ],
        [0.33333333, 0.94045574, 0.        , 0.        , 0.        ],
        [1.        , 2.15470054, 3.20626738, 0.        , 0.        ],
        [3.        , 4.46410162, 5.30940108, 5.82260132, 0.        ],
        [9.        , 7.39230485, 7.        , 6.84900179, 6.780246  ]]))

Solution: $P_4(\sqrt{3}) = 6.78024599976228$.

#### (b) $f(x) = \sqrt{x}$ and the values $x_0 = 0, x_1 = 1, x_2 = 2, x_3 = 4$, and $x_4 = 5$.

In [40]:
def g(x):
    return math.sqrt(x)

neville(math.sqrt(3),[0,1,2,4,5],[g(0),g(1),g(2),g(4),g(5)])

(1.3303367684185772,
 array([[0.        , 0.        , 0.        , 0.        , 0.        ],
        [1.        , 1.73205081, 0.        , 0.        , 0.        ],
        [1.41421356, 1.30322537, 1.36067709, 0.        , 0.        ],
        [2.        , 1.33573306, 1.3111578 , 1.33923461, 0.        ],
        [2.23606798, 1.46460982, 1.32422225, 1.31354876, 1.33033677]]))

Solution: $P_4(\sqrt{3}) = 1.3303367684185772$.

#### (c) Compare the accuracy of the approximation in parts (a) and (b).

In [41]:
a = abs(3**math.sqrt(3) - 6.78024599976228)
b = abs(math.sqrt(math.sqrt(3)) - 1.3303367684185772)
print(a,b)

0.0752541459364009 0.014262755466084798


We see that the accuracy is better for (b).

#### 6. Neville’s method is used to approximate $f (0.5)$, giving the following table. Determine $P_2 = f(0.7)$.

\begin{align*}
P_{0,1,2} = \dfrac{27}{7} &= \dfrac{(x-x_0)P_{1,2} - (x-x_2)P_{0,1}}{x_2-x_0} \\
&= \dfrac{(0.5-0)P_{1,2} - (0.5-0.75)3.5}{0.7} \\
2.7 &= 0.5P_{1,2} + (0.2)3.5 \\
2 &= 0.5P_{1,2} \\
4 &= P_{1,2} \\
&= \dfrac{(x-x_1)P_{2} - (x-x_2)P_{1}}{x_2-x_1} \\
&= \dfrac{0.5-0.4(P_2) - (0.5-0.7)2.8}{0.7-0.4} \\
1.2 &= 0.1P_2 + 0.56 \\
0.64 &= 0.1P_2 \\
P_2 &= 6.4.
\end{align*}

#### 12. Use iterated inverse interpolation to find an approximation to the solution of $x - e^{-x} = 0$, using the data.

In [42]:
neville(0,[0.740818,0.670320,0.606531,0.548812],[0.3,0.4,0.5,0.6])

(2.288172726124393,
 array([[0.3       , 0.        , 0.        , 0.        ],
        [0.4       , 1.35083548, 0.        , 0.        ],
        [0.5       , 1.45083949, 1.90252526, 0.        ],
        [0.6       , 1.55083421, 2.00247778, 2.28817273]]))

One solution would be $x = 2.288172726124393$.

## Section 3.3